src/three/plugins/SUBTREELoader.js

```diff
@@ -0,0 +1,730 @@
+				tileAvailabilityBits
+			);
+			subtree._contentAvailabilityBitstreams.push(bitstream);
+		}
+
+		subtree._childSubtreeAvailability = this.parseAvailabilityBitstream(
+			subtreeJson.childSubtreeAvailability,
+			bufferViewsU8,
+			childSubtreeBits
+		);
+
+	}
+
+	/**
+	 * A helper object for storing the two parts of the subtree binary
+	 *
+	 * @typedef {object} ParsedBitstream
+	 * @property {Boolean} constant
+	 * @property {ArrayBuffer} bitstream
+	 * @property {number} lengthBits The length of the availability bitstream in bits
+	 * @private
+	 */
+
+
+	/**
+	 * Given the JSON describing an availability bitstream, turn it into an
+	 * in-memory representation using an object. This handles bitstreams from a bufferView.
+	 *
+	 * @param {Object} availabilityJson A JSON object representing the availability
+	 * @param {Object} bufferViewsU8 A dictionary of bufferView index to its Uint8Array contents.
+	 * @param {number} lengthBits The length of the availability bitstream in bits
+	 * @returns {ParsedBitstream}
+	 * @private
+	 */
+	parseAvailabilityBitstream(
+		availabilityJson,
+		bufferViewsU8,
+		lengthBits,
+	) {
+
+		if (!isNaN(availabilityJson.constant)) {
+			return {
+				constant: Boolean(availabilityJson.constant),
+				lengthBits: lengthBits,
+			};
+		}
+		let bufferView;
+		// Check for bitstream first, which is part of the current schema.
+		// bufferView is the name of the bitstream from an older schema.
+
+		if (!isNaN(availabilityJson.bitstream)) {
+			bufferView = bufferViewsU8[availabilityJson.bitstream];
+		} else if (!isNaN(availabilityJson.bufferView)) {
+			bufferView = bufferViewsU8[availabilityJson.bufferView];
+		}
+		return {
+			bitstream: bufferView,
+			lengthBits: lengthBits
+		};
+
+	}
+
+
+	/**
+	 * Expand a single subtree tile. This transcodes the subtree into
+	 * a tree of {@link SubtreeTile}. The root of this tree is stored in
+	 * the placeholder tile's children array. This method also creates
+	 * tiles for the child subtrees to be lazily expanded as needed.
+	 *
+	 * @param {Object | SubtreeTile} subtreeRoot The first node of the subtree
+	 * @param {Subtree} subtree The parsed subtree
+	 * @private
+	 */
+	expandSubtree(subtreeRoot, subtree) {
+
+		// If the subtree root tile has content, then create a placeholder child with cloned parameters
+		// Todo Multiple contents not handled, keep the first found content
+		for (let i = 0; subtree && i < subtree._contentAvailabilityBitstreams.length; i++) {
+			if (subtree && this.getBit(subtree._contentAvailabilityBitstreams[i], 0)) {
+
+				// Create a child holding the content uri, this child is similar to its parent and doesn't have any children
+				let contentTile = {
+					content: {uri: this.parseImplicitURI(subtreeRoot, this.rootTile.__contentUri)},
+					boundingVolume: subtreeRoot.boundingVolume,
+					geometricError: subtreeRoot.geometricError
+				}
+				subtreeRoot.children.push(contentTile);
+				break;
+			}
+		}
```

Comment:
Can you explain this? It looks like this child is added in order to support geometry for the level 0 tile but I'm not seeing how the next level children are added to this tile. From [transcodeSubtreeTiles](https://github.com/NASA-AMMOS/3DTilesRendererJS/pull/650/files#diff-e1ca0fc4e2fdac866a35e569dee4e3e37a9545e4fd24c395eebbd5ca998a0b1cR489) it looks like all the next level 1 children are all also added as children to the subtree root meaning the level 0 tile and level 1 tiles would be siblings. Is that correct? In order for the structure of the tile set to be correct and for tile traversal to work correctly the level 1 tiles need to be children of the level 0 content tile.

It's okay to _always_ add another level 0 content tile (even if it's empty) here and then expand the children into that to handle this case.

Reply:
You're correct, it's exactly what happens here.
I add the level 0 tile with content as a children and I add all the children to the "original" level 0 tile (the one with subtree uri). So as you said the level 0 tile will be siblings with the children tiles :/
I cannot replace the level 0 tile subtree by the level 0 tile content, they both need to exist, we agree on this.

> It's okay to always add another level 0 content tile (even if it's empty) here and then expand the children into that to handle this case.

1. You mean I should use the one I've just created and call `transcodeSubtreeTiles` with it ? Like this:

```js
let contentTile;
for ( let i = 0; subtree && i < subtree._contentAvailabilityBitstreams.length; i ++ ) {

	if ( subtree && this.getBit( subtree._contentAvailabilityBitstreams[ i ], 0 ) ) {

		// Create a child holding the content uri, this child is similar to its parent and doesn't have any children
		contentTile = {
			content: { uri: this.parseImplicitURI( subtreeRoot, this.rootTile.__contentUri ) },
			boundingVolume: subtreeRoot.boundingVolume,
			geometricError: subtreeRoot.geometricError
		};
		subtreeRoot.children.push( contentTile );
		break;

	}

}


// Creating each leaf inside the current subtree
const bottomRow = this.transcodeSubtreeTiles(
	contentTile || subtreeRoot,  //HERE I'M USING THE NEW TILE AS PARENT
	subtree
);

```

2. Or I should add it the parentTile:

```js

for ( let i = 0; subtree && i < subtree._contentAvailabilityBitstreams.length; i ++ ) {

	if ( subtree && this.getBit( subtree._contentAvailabilityBitstreams[ i ], 0 ) ) {

		// Create a child holding the content uri, this child is similar to its parent and doesn't have any children
		const contentTile = {
			content: { uri: this.parseImplicitURI( subtreeRoot, this.rootTile.__contentUri ) },
			boundingVolume: subtreeRoot.boundingVolume,
			geometricError: subtreeRoot.geometricError
		};
		subtreeRoot.parent.children.push( contentTile ); // HERE I add it to the parent
		break;

	}

}

```

3. Or I combine both ?


The last one (`subtreeRoot.parent.children.push`) was causing issue with the break the `__depth` condition on the child in `ensureChildrenArePreprocessed`.
But if we replace `break` by `continue`, there is no issue (cf https://github.com/NASA-AMMOS/3DTilesRendererJS/pull/650#issuecomment-2283609240) . Except in the case of a subtree root tile who also has content, because it won't have parent... So maybe I should just keep the 1st implem (use `contentTile` as parent for the other children)

Reply:
My first implementation isn't completely correct though.
As you said, I should always add the empty level 0 tile but to make it work I need to have some attributes.
I don't see how to handle this properly. I need to copy the attributes from the original tile know that the SubtreeTile class did bother you.
To make it more comprehensible, you can check my latest commit 87e9c3193677c14f0a51e39042fbd12ed6576112
This static function doesn't satisfy me

Reply:
It's okay - it could just be a function that returns the cloned element but the static functions works. The only thing that sticks out to me is this line:

```
copyTile.parent = tile.parent;
```

The "copied tiles" parent should not be set to the same as the parent of it's _actual_ parent (the tile it's a child of). This will get overwritten in preprocess node, though, so there's no reason to set the parent here or in the SubTreeTile constructor.

---

Ultimately this is an artifact of the fact that "multiple contents" isn't support right now. If multiple contents were supported then this tile could contain both renderable and un renderable content. We can reevaluate this in the future. Maybe worth just adding a comment inline to note that.